### ISSUE TITLE
Pass back a future's result in the `ClosureTable`

### ISSUE THREAD

**Issue by ggreif** [dfinity/motoko#2679]
This is an optimisation follow-up to #2663.

**Comment by crusso** [comment]
You are right about the change in semantics (since we skip the (de)serialization step). I wonder if it would make sense to delay this until we actually refine the type system so that the contents of an async type need not be shared per se, but an async type is shared (and useable as a shared method return) only when its contents is shared. Joachim and I had discussed doing this in the past to generalize what you can return in async blocks.
